src/middlewared/middlewared/plugins/disk_/format.py

```diff
@@ -28,46 +25,58 @@ def validate_disk(self, disk, swapgb):
-        # partition starts at multiple of 128
-        swapsize = int(((swapsize / sectorsize) + 127) / 128) * 128
-
-        return swapsize, alignment
-
     @private
-    def format(self, disk, swapgb, sync=True):
-        swapsize, alignment = self.validate_disk(disk, swapgb)
+    def format(self, disk, swapgb):
+        self.validate_disk(disk, swapgb)
 
-        job = self.middleware.call_sync('disk.wipe', disk, 'QUICK', sync)
+        job = self.middleware.call_sync('disk.wipe', disk, 'QUICK', False)
         job.wait_sync()
         if job.error:
             raise CallError(f'Failed to wipe disk {disk}: {job.error}')
 
-        if swapsize > 0:
-            commands = [
-                ('sgdisk', f'-a{alignment}', f'-n1:128:{swapsize}', f'-t1:{SWAP_PARTHEX}', f'/dev/{disk}'),
-                ('sgdisk', '-n2:0:0', f'-t2:{ZFS_PARTHEX}', f'/dev/{disk}'),
-            ]
+        device = parted.getDevice(f'/dev/{disk}')
+        device.clobber()
+        parted_disk = parted.freshDisk(device, 'gpt')
+
+        if swapgb > 0:
+            swap_geometry = parted.Geometry(
+                start=0,
+                length=parted.sizeToSectors(swapgb, 'GiB', device.sectorSize),
+                device=device,
+            )
+            swap_filesystem = parted.FileSystem(type='linux-swap(v1)', geometry=swap_geometry)
+            swap_partition = parted.Partition(
+                disk=parted_disk,
+                type=parted.PARTITION_NORMAL,
+                geometry=swap_geometry,
+                fs=swap_filesystem,
+            )
+            parted_disk.addPartition(swap_partition, constraint=device.optimalAlignedConstraint)
+
+            data_start = swap_partition.geometry.end
         else:
-            commands = [('sgdisk', f'-a{alignment}', '-n1:0:0', f'-t1:{ZFS_PARTHEX}', f'/dev/{disk}')]
+            data_start = 0
+
+        data_geometry = parted.Geometry(
+            start=data_start,
+            end=device.length,
+            device=device
+        )
+        data_filesystem = parted.FileSystem(type='zfs', geometry=data_geometry)
+        data_partition = parted.Partition(
+            disk=parted_disk,
+            type=parted.PARTITION_NORMAL,
+            geometry=data_geometry,
+            fs=data_filesystem,
+        )
+        parted_disk.addPartition(data_partition, constraint=device.optimalAlignedConstraint)
+        parted_disk.commit()
 
         # TODO: Install a dummy boot block so system gives meaningful message if booting from a zpool data disk.
 
-        for cmd in commands:
-            cp = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
-            if cp.returncode != 0:
-                raise CallError(f'Unable to GPT format the disk "{disk}": {cp.stderr}')
-
         self.middleware.call_sync('device.settle_udev_events')
 
         for partition in self.middleware.call_sync('disk.list_partitions', disk):
```

Comment:
running `disk.list_partitons` for every disk and then calling `zfs.pool.clear_label` is unnecessary because `disk.wipe, QUICK` wipes the first and last 32MB of the disk.

Reply:
Why is this code here then?
And what happens if we had a swap + partition; in that case, `disk.wipe` will only be able to erase partition end, but won't affect partition start?

Reply:
We wipe the first and last 32MB of a disk regardless of what is on the disk in `disk.wipe`? Also, when I was investigating this, `device.clobber()` removes partition tables off the disk so I'm not even sure we need the call to `disk.wipe`. I can't remember if `device.clobber()` removed table entries at end of disk.

Reply:
@sonicaj you introduced this in e0d120f4284d139c989b8776452617a57272c79f - why?

Reply:
https://ixsystems.atlassian.net/browse/NAS-103069 that was the original issue @yocalebo

Reply:
I understand but that change was made before we wiped the first and last 32MB of a disk IIRC which means the zfs partition label will always be removed.

Also, have you confirmed that the `clobber()` method removes all partition labels off the disk?